src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs

```diff
@@ -486,61 +469,662 @@ .maxstack  2
+            compVerifier.VerifyIL("C.M<T, U>(U)",
+@"{
+  // Code size       35 (0x23)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  box        ""U""
+  IL_0006:  isinst     ""T""
+  IL_000b:  brfalse.s  IL_0022
+  IL_000d:  ldarg.0
+  IL_000e:  box        ""U""
+  IL_0013:  unbox.any  ""T""
+  IL_0018:  box        ""T""
+  IL_001d:  call       ""void System.Console.Write(object)""
+  IL_0022:  ret
+}");
+        }
+
+        [Fact, WorkItem(20641, "https://github.com/dotnet/roslyn/issues/20641")]
+        public void PatternsVsAs01()
+        {
+            var source = @"using System.Collections;
+using System.Collections.Generic;
+
+class Program
+{
+    static void Main() { }
+
+    internal static bool TryGetCount1<T>(IEnumerable<T> source, out int count)
+    {
+        ICollection nonGeneric = source as ICollection;
+        if (nonGeneric != null)
+        {
+            count = nonGeneric.Count;
+            return true;
+        }
+
+        ICollection<T> generic = source as ICollection<T>;
+        if (generic != null)
+        {
+            count = generic.Count;
+            return true;
+        }
+
+        count = -1;
+        return false;
+    }
+
+    internal static bool TryGetCount2<T>(IEnumerable<T> source, out int count)
+    {
+        switch (source)
+        {
+            case ICollection nonGeneric:
+                count = nonGeneric.Count;
+                return true;
+
+            case ICollection<T> generic:
+                count = generic.Count;
+                return true;
+
+            default:
+                count = -1;
+                return false;
+        }
+    }
+}";
+            var compilation = CreateCompilation(source, options: TestOptions.ReleaseDll);
+            compilation.VerifyDiagnostics();
+            var compVerifier = CompileAndVerify(compilation);
+            compVerifier.VerifyIL("Program.TryGetCount1<T>",
+@"{
+  // Code size       45 (0x2d)
+  .maxstack  2
+  .locals init (System.Collections.ICollection V_0, //nonGeneric
+                System.Collections.Generic.ICollection<T> V_1) //generic
+  IL_0000:  ldarg.0
+  IL_0001:  isinst     ""System.Collections.ICollection""
+  IL_0006:  stloc.0
+  IL_0007:  ldloc.0
+  IL_0008:  brfalse.s  IL_0014
+  IL_000a:  ldarg.1
+  IL_000b:  ldloc.0
+  IL_000c:  callvirt   ""int System.Collections.ICollection.Count.get""
+  IL_0011:  stind.i4
+  IL_0012:  ldc.i4.1
+  IL_0013:  ret
+  IL_0014:  ldarg.0
+  IL_0015:  isinst     ""System.Collections.Generic.ICollection<T>""
+  IL_001a:  stloc.1
+  IL_001b:  ldloc.1
+  IL_001c:  brfalse.s  IL_0028
+  IL_001e:  ldarg.1
+  IL_001f:  ldloc.1
+  IL_0020:  callvirt   ""int System.Collections.Generic.ICollection<T>.Count.get""
+  IL_0025:  stind.i4
+  IL_0026:  ldc.i4.1
+  IL_0027:  ret
+  IL_0028:  ldarg.1
+  IL_0029:  ldc.i4.m1
+  IL_002a:  stind.i4
+  IL_002b:  ldc.i4.0
+  IL_002c:  ret
+}");
+            compVerifier.VerifyIL("Program.TryGetCount2<T>",
+@"{
+  // Code size       47 (0x2f)
```

Comment:
This is the driving use case for this optimization. See #20641. Now the code is nearly the same as the hand-written alternative, with the exception of a different ordering of the basic blocks which causes a couple of extra IL bytes.